spotify_tensorflow/tfx/tft.py

```diff
@@ -0,0 +1,214 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2017 Spotify AB.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from __future__ import absolute_import, division, print_function
+
+import argparse
+import os
+import sys
+from abc import abstractmethod, ABCMeta
+from typing import Dict, Union, Any, List  # noqa: F401
+
+import apache_beam as beam
+import six
+from apache_beam.io import tfrecordio
+from apache_beam.io.filesystem import CompressionTypes
+from apache_beam.runners import PipelineState  # noqa: F401
+from spotify_tensorflow.tf_schema_utils import schema_txt_to_feature_spec
+from spotify_tensorflow.tfx.utils import assert_not_empty_string, assert_not_none
+from tensorflow_transform.beam import impl as beam_impl
+from tensorflow_transform.beam.tft_beam_io import transform_fn_io
+from tensorflow_transform.coders import ExampleProtoCoder
+from tensorflow_transform.tf_metadata import dataset_metadata
+from tensorflow_transform.tf_metadata import dataset_schema
+
+
+@six.add_metaclass(ABCMeta)
+class TFTransform:
```

Comment:
So just want to make sure I get this straight - the way this works is that `TfTransformTask` expects the user to provide a python script location, and that script should have a class that extends this providing preprocessing_function, right?

Reply:
I would prefer that, instead of making users extend a class to use the lib, if we can instead make this a concrete class with a constructor that takes `preprocessing_fn` as an argument? Might be a bit nicer from the user perspective, what do you think? 

Also - I wonder if it would be cleaner + safer (can enforce type checks) if we could allow users to define preprocessing_fn directly in the luigi task...

Reply:
Hmm, I'm not sure how dynamically passing `preprocessing_fn` as an argument would work, since the script which contains `preprocessing_fn` will be sent to workers and get executed. It may work if we ask users to define the function in luigi job and somehow auto generate this tft dataflow script, but it seems an overkill to me and it may require some dark magic.

The other way is to use the luigi script also as the tft dataflow job script, but we'll be mixing luigi orchestration stuff with the actual tft logic. A separate executable python script for tft would be nice to have from open source point of view or in case we are using a different orchestration framework in the future. 

But I think it's an interesting point, I want to make sure I understand it correctly. Let's chat more tomorrow in person.

Reply:
Ah, totally see the point now - my mistake! Definitely makes sense to separate orchestration logic and pipeline logic 👍 I guess my only (very minor) concern now is just that users have to extend this class (from a library) in their script, IMO it's a bit of a smell, although not a super strong opinion or anythin. Let's chat about this in person